src/graph.ts

```diff
@@ -1,30 +1,88 @@
-import { PackageInfos } from "./types/PackageInfo";
-import { getInternalDeps } from "./dependencies";
+import type { PackageInfos } from "./types/PackageInfo";
+import type { DependencyMap } from "./createDependencyMap";
+import type { PackageGraph, PackageGraphVisitor } from "./types/PackageGraph";
 
-const graphCache = new Map<PackageInfos, [string, string][]>();
+import { createDependencyMap } from "./createDependencyMap";
+import multimatch from "multimatch";
 
-export function getPackageGraph(packages: PackageInfos) {
-  if (graphCache.has(packages)) {
-    return graphCache.get(packages)!;
-  }
+// Reference: https://github.com/pnpm/pnpm/blob/597047fc056dd25b83638a9ab3df0df1c555ee49/packages/filter-workspace-packages/src/parsePackageSelector.ts
+export interface PackageGraphScope {
+  namePatterns?: string[];
+  includeDependencies?: boolean;
+  includeDependents?: boolean;
+}
+
+export function createPackageGraph(packages: PackageInfos, scope: PackageGraphScope = {}): PackageGraph {
+  const dependencyMap = createDependencyMap(packages);
 
-  const edges: [string, string][] = [];
+  const packageSet = new Set<string>();
+  const edges: PackageGraph["dependencies"] = [];
 
-  for (const [pkg, info] of Object.entries(packages)) {
-    const deps = getInternalDeps(info, packages);
-    for (const dep of deps) {
-      edges.push([dep, pkg]);
+  const visitor: PackageGraphVisitor = (pkg, dependencies, dependents) => {
```

Comment:
The visitor operates on the assumption that all packages that are visited are relevant to the graph. It doesn't impact the traversal, it only affects the output.

It will include the current node's dependencies and dependents based on the values of the input flags.